### PR TITLE
decentralized-security-marketplace escrow protocol proposal

### PR DESCRIPTION
**Pull request by aka-toxa**
#### Project Abstract

This is a response for a https://github.com/w3f/Grants-Program/pull/1452 RFP

This project aims to build a universal escrow protocol on !ink to let other developer build their own marketplaces or manage their payments in transparent and fair way. This project derives an idea of decentralised security marketplace proposed by the grants team, so the use-case described below respect an idea about security audits. However this project may be used in many other ways.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [X] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [X] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [X] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [X] The software delivered for this grant will be released under an open-source license specified in the application.
- [X] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

File: applications/decentralized-security-marketplace.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
+# Decentralized Security Marketplace
+
+
+- **Team Name:** Magic Powered Inc,
+- **Payment Address:** Magicpowered.eth, USDC, Ethereum
```
Comment (0xCaso) [inline applications/decentralized-security-marketplace.md:5]:
```suggestion
- **Payment Address:** Magicpowered.eth (0x85f9c38a44EfB45CeF47cBf510e6e18cDdf2a78A), USDC, Ethereum
```

File: applications/decentralized-security-marketplace.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+Judges then review all report submissions and verify them. After validation, they vote for the proposal. When the consensus of judges accepts the proposal, it executes by sending rewards to the engineers.
+
+![Successfull report](https://github.com/magic-powered/decentralised-security-marketplace-architecture/blob/main/img/successfull-report.png?raw=true)
+
+The judge can change the reward size in the proposal and the serenity of found issues at any time by consensus mechanism together with other judges.
+
+![Change report submission](https://github.com/magic-powered/decentralised-security-marketplace-architecture/blob/main/img/change-proposal.png?raw=true)
+
+Any eligible community member can become a judge. Judges are rewards from the same rewards pool with a percent of rewards to be distributed by the proposals they review.
```
Comment (0xCaso) [inline applications/decentralized-security-marketplace.md:57]:
```suggestion
Any eligible community member can become a judge. Judges are rewarded from the same rewards pool with a percent of rewards to be distributed by the proposals they review.
```

File: applications/decentralized-security-marketplace.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+| **0a.** | License                   | MIT                                                                                                                                                                     |
+| **0b.** | Documentation             | Both inline and markdown documentation pages describing the contract functions and other necessary info: How to run, contract composition diagrams, how to call, etc... |
+| **0c.** | Testing and Testing Guide | Unit tests and functional tests for the contracts                                                                                                                       |
+| **0d.** | Code repo                 | Repo with the code of smart contracts                                                                                                                                   |
+|  **1.** | Deployment                | Smart contracts deployed to mainnet environment                                                                                                                         |
```
Comment (0xCaso) [inline applications/decentralized-security-marketplace.md:191]:
```suggestion
| **1.** | Code repo                 | Repo with the code of smart contracts                                                                                                                                   |
|  **2.** | Deployment                | Smart contracts deployed to mainnet environment                                                                                                                         |
```

**Review by 0xCaso — CHANGES_REQUESTED** [review]
I also added some mini fixes

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application and your interest in our grant program. I also shared the application with the people who wrote the RFP, but apart from I have a few initial questions/comments: 

- We usually don’t support the deployment of a smart contract. Only the technical development and the work that is useful for others. (e.g. Smart Contracts version 1.0) or milestone 4 deployment. One solution might be, if you don’t have any other business model, to initially apply for a grant to develop a PoC and after this to apply for treasury funding. 
- Usually, we ask teams to initially provide some mock-ups or designs for front-end focused applications or at least parts of them.

**Comment by bhargavbh** [comment]
Thank your for your interest. The proposal seems interesting but i have a few high-level questions before getting into technical aspects:

1. The way the process is currently setup, Judges seem to be playing an extremely central role (more than necessary in my opinion). For most audit reports submitted by the engineers, the projects (who requested for audit) are in the best position to evaluate the audit report. If the engineer and project mutually agree with the findings of audit, then there should be a short-circuit to directly approve the fund-release without involving judges. It is only when there is a dispute between the engineer and project, the judge should be involved. 
2. what is the mechanism you propose to keep the judges anonymous. This is to avoid collusion between the parties. For e.g., if the engineer can find out who the judge is, then he can bribe to get the audit passed. Anonymity is crucial for the whole system to work and some core concepts like VRF can help in achieving it, demonstrating a clear delta over Web2 solutions. 
3. What substrate specific technologies do you plan to use. By 'Rust smart contracts' do you mean ink! ?
4. How modular can this solution be implemented so that the core template can be used for any marketplace supported by escrows?

**Comment by aka-toxa** [comment]
Hey @Noc2 


> * We usually don’t support the deployment of a smart contract. Only the technical development and the work that is useful for others. (e.g. Smart Contracts version 1.0) or milestone 4 deployment. One solution might be, if you don’t have any other business model, to initially apply for a grant to develop a PoC and after this to apply for treasury funding.

well, the core of this project is a transparent escrow protocol it is essential and I don't see any ways to extract or remove it from the project. Since the whole thing will be open-sourced I see a huge benefit for the community for both frontend and protocol implementation. 

I see your point splitting proposals into two: PoC (protocol, minimal frontend) and then Product itself, let me discuss it with my team

> * Usually, we ask teams to initially provide some mock-ups or designs for front-end focused applications or at least parts of them.

I'm not sure it is a front-end focused application though. We are looking to implement the reputation-based escrow protocol where the community can rely on reputation and prev. work of others to be able to solve their security issues. I'd say the frontend here is secondary and at the same time important

**Comment by aka-toxa** [comment]
hey @bhargavbh 

> * The way the process is currently setup, Judges seem to be playing an extremely central role (more than necessary in my opinion). For most audit reports submitted by the engineers, the projects (who requested for audit) are in the best position to evaluate the audit report. If the engineer and project mutually agree with the findings of audit, then there should be a short-circuit to directly approve the fund-release without involving judges. It is only when there is a dispute between the engineer and project, the judge should be involved.

yeah some part of my team think the same way. I think we can provide flexibility: project with judges or without. 
My personal experience working with my customers tells me that most of the projects cannot verify the work of security engineers by them selfs and they could ask for thrid-party judges just to help verify findings, so they could be involved even if there no dispute but expertise of the project team is not enough to accept the work. 

But ultimately this system does not require thrid party judges, projects can approve report as well, I will add this note to the proposal...

> * what is the mechanism you propose to keep the judges anonymous. This is to avoid collusion between the parties. For e.g., if the engineer can find out who the judge is, then he can bribe to get the audit passed. Anonymity is crucial for the whole system to work and some core concepts like VRF can help in achieving it, demonstrating a clear delta over Web2 solutions.

I don't think keeping judges anonymous is a good idea, because every time you have anonymous judges there is always corruption happens. Having fully transparent and visible mechanisms helps this system build trust and reputation for every party involved: Judges, Engineers, Projects. Every Judge, Engineer and Project will have a transparent report history, proposal history, judgements history, disputes history and it will be very easy define who is fair and who is not.

> * What substrate specific technologies do you plan to use. By 'Rust smart contracts' do you mean ink! ?

Yes! I mean ink! but some part of my team thinking implementing the whole protocol on solidity and deploy to evm compatible parachain but I think it would be better to stick with something that any parachain can run so anyone can deploy the protocol anywhere, what do you think?

> * How modular can this solution be implemented so that the core template can be used for any marketplace supported by escrows?

Yes, the idea is that the protocol can be used anywhere for any purposes where escrow can be applied. This is actually a good question that help me see that probably @Noc2 was right and it is better to start from the protocol first do not think about the security marketplace for now and then let everyone build any type of marketplace they want

**Comment by 0xCaso** [comment]
> Yes! Issues could be found in the projects are categorised by serenity usually, for example: low, medium, critical
> 
> each issue type can have it's own "price". For example, Project announce it will pay for each low serenity issue 50 USD, for medium 150 USD and for critical 500 USD
> 
> Every time when Engineer ready to submit report they estimate issues serenity by them selfs and put it into the proposal. For example, if I found 2 low serenity issues and 1 critical issue, I put my report to the IPFS, and specify this in the proposal. The protocol can calculate 2 * 50 + 1 * 500 = 600 USD and put it as a value of the proposal. When everyone agree with that and everything is fine - proposal is approved and 600 USD goes to the proposal author

Ok clear, thanks, that makes sense and should be feasible to implement.

> Yes we are thinking about this. Ideally we would like to see a lot of community members as Judges and Ideally they would have to manually apply to the project to be a Judge for this particular project. All voting and all activities should be transparent and clear for the whole community to ensure fairness and good reputation for all actors

Yes, that's also a nice solution to make Judges apply for the single project. Maybe it would be nice to track Judge in a dashboard, so it's visible how many times they performed and misbehaved.
Btw, I also agree with the point 1. of @bhargavbh's comment.

> Thats a good point, we have dispute in our team right now about that as well I might update the proposal with something we come up with that would be better than AWS, thanks!

Great!

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. A few follow-up comments:

- Regarding the front end, at the moment, we would basically only really see the scope of the second milestone (19k) after the delivery of the first milestone, “Frontend app figma design”. This is partly why for front-end applications, we require some initial designs/mock-ups upfront. But feel free to remove the front end entirely for now. 
- In general, the deliveries of the milestones need to be better defined. The milestone tables are basically the requirements of our contracts, and for example, “Repo with the code of smart contracts” could mean anything at the moment. It would be nice to have here the functionality of the smart contract described as well as the programming language, etc. I assume you want to leverage ink! For this project. Also, for Souldbound NFT, I believe there are already some smart contracts available that you could potentially reuse. 
- Feel free to remove the deployment from the deliveries.

**Comment by bhargavbh** [comment]
Thanks for the replies. 
> yeah some part of my team think the same way. I think we can provide flexibility: project with judges or without. My personal experience working with my customers tells me that most of the projects cannot verify the work of security engineers by them selfs and they could ask for thrid-party judges just to help verify findings, so they could be involved even if there no dispute but expertise of the project team is not enough to accept the work.
> But ultimately this system does not require thrid party judges, projects can approve report as well, I will add this note to the proposal...

This is subjective, however it should not be hard to change the underlying platform based on user data and feedback at a later stage. I would suggest, both methods be implemented (i.e., always via judge and short-circuiting)

> I don't think keeping judges anonymous is a good idea, because every time you have anonymous judges there is always corruption happens. Having fully transparent and visible mechanisms helps this system build trust and reputation for every party involved: Judges, Engineers, Projects. Every Judge, Engineer and Project will have a transparent report history, proposal history, judgements history, disputes history and it will be very easy define who is fair and who is not.

Not sure if reputation system is the best approach (specially for audit services), with all the possible sybil attack surfaces opening up. Since there is no verification of physical identity, banned actors can always rejoin with new accounts making the reputation system ineffective. I however do see you point that one-way anonymity is again ineffective. I think there is a huge design space to be explored here with several considerations that are only clear during the course of implementation. I guess the guiding principles that we can agree upon at this stage are sybil resistance and corruption resistance to a reasonable degree. 
> Yes! I mean ink! but some part of my team thinking implementing the whole protocol on solidity and deploy to evm compatible parachain but I think it would be better to stick with something that any parachain can run so anyone can deploy the protocol anywhere, what do you think?

ink! contracts are something that we are happy to support, and since there is no escrow-service based template already existing in ink, it could be the core contribution of this project. 

> Yes, the idea is that the protocol can be used anywhere for any purposes where escrow can be applied. This is actually a good question that help me see that probably @Noc2 was right and it is better to start from the protocol first do not think about the security marketplace for now and then let everyone build any type of marketplace they want. 

Great, it would be ideal to have a modular escrow template and derive the specific auditing platform contract from it. 

In general, as suggested by @Noc2 , please add more details in the deliverables section and structure each deliverable into subtasks if needed.

**Comment by aka-toxa** [comment]
Thank you guys for all your replies! 
It is indeed better for the community to have opensource escrow protocol then just a single marketplace 

I'm now preparing contract-focused application with a bit more details about the protocol and contracts collected last week with my technical design team. what do you think would that be better to close this application and start new pull request with the link to this one or I can just edit this one?

**Comment by 0xCaso** [comment]
> what do you think would that be better to close this application and start new pull request with the link to this one or I can just edit this one?

It would be good to edit this one, if you don't have to completely change the application. Thanks!

**Comment by aka-toxa** [comment]
hey team 
I've updated application, removed all frontend work and focused on the escrow protocol. 

I've provided dirty and hacky interfaces for smart contracts (those a not real interfaces, it just to show an idea and move my answers to your question inside the application) 

We are still debating about voting policies and how to apply them. I'm proposing to go with just an enum with three predefined rules: **admin only** - when proposals can be closed by just project admin (no judges), **judges only** - when proposals can be closed only by judges, **both** - when both admin and judges should vote to close a proposal. I think these three should be sufficient for escrow protocol, since we are not doing complex DAO. wdyt? 

Please let me know if you also have further questions.

**Comment by aka-toxa** [comment]
hey team, any further comments here?

**Review by Noc2 — CHANGES_REQUESTED** [review]
I'm sorry for not getting back to you sooner, and thanks for the update. I have a couple of follow-up questions:

- What is the reason for developing the “Smart contract API Mock” as the first milestone? 
- Could you please let me know why you removed the tests and docker for the smart contracts?
- In general, it would be nice to have the technical details, like the programming language ink!, as part of the milestone tables since these are the requirements of our contracts. 
- Apart from that, the application seems relatively expensive to me at the moment, but this is just my personal opinion.

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application. I would love to see this happen. I added a couple of questions/comments below. But the main request I have is to add a report or otherwise thorough documentation to M1. I think the main reason why the first step in the RFP is "research and design" is because we want the framework to be both robust _and_ flexible enough to accommodate all kinds of marketplaces. I would be interested in hearing about the reasoning for the (eventually) chosen design. Ideally, we would discuss and dissect it during M1 review. As a simple interface implementation, M1 doesn't make sense. I think this is what @Noc2 is alluding to.

**Comment by semuelle** [comment]
pinging @aka-toxa

**Comment by aka-toxa** [comment]
sorry guys I was at unexpected sick leave last week :( 
everything fine by now. going to address all comments now

**Comment by aka-toxa** [comment]
hey @Noc2 answering to your comments

> I'm sorry for not getting back to you sooner, and thanks for the update. I have a couple of follow-up questions:
> 
> What is the reason for developing the “Smart contract API Mock” as the first milestone?

Good call out, it was something that I wasn't cleaned properly

> Could you please let me know why you removed the tests and docker for the smart contracts?

Tests will be implemented before smart contract implementation at Milestone 1. let me rephrase deliverable to be more explicit there. 

My team really loves TDD approach in development, where we know exactly what we build and we already have large test coverage and then we do specific implementation.

> In general, it would be nice to have the technical details, like the programming language ink!, as part of the milestone tables since these are the requirements of our contracts.

Sure thing I will add it!

File: applications/decentralized-security-marketplace.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+| **0a.** | License                   | MIT                                                                                                                                                                                                                                                                                      |
+| **0b.** | Documentation             | Inline code documentation for public interface                                                                                                                                                                                                                                           |
+| **0c.** | Testing and Testing Guide | Tests for each public function on the smart contract                                                                                                                                                                                                                                     |
```
[the remark this answers is not in the document]
Comment (aka-toxa) [reply applications/decentralized-security-marketplace.md:233]: yes! my team really likes TDD approach in any type of development. Of course it won't limit us and we will extend tests while implementing. Let me rephrase all of this to be more explicit

File: applications/decentralized-security-marketplace.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+- **Estimated Duration:** 1 month
+- **FTE:**  1.5 
+- **Costs:** 22,000 USD
+
+
+|   Number | Deliverable                 | Specification                                                                                                                                                               |
+|---------:|-----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+|  **0a.** | License                     | MIT                                                                                                                                                                         |
+|  **0b.** | Documentation               | Markdown documents describing user flow and details on how to run RewardsPool contract and call it from the cli                                                             |
+|  **0c.** | Testing and Testing Guide   | N/A                                                                                                                                                                         |
```
[the remark this answers is not in the document]
Comment (aka-toxa) [reply applications/decentralized-security-marketplace.md:248]: because intention was to write most of the tests before implementation. the M1 should be very heavy loaded design and initial TDD phase that will give my team understanding what exactly they need to implement and how. 

I expect some additional tests in M2/M3 anyway, so let me add them here. But the main intention that the most coverage will be done in M1

File: applications/decentralized-security-marketplace.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    
+    impl RewardsPool {
+        /// create an empty project. Recently created project will have a status: Pending
+        pub fn create_project(&mut self, id: ProjectId) -> ProjectId {}
+        /// defines whatever who can vote for rewards proposals
+        pub fn set_project_policy(&mut self, id: ProjectId, policy: ProjectPolicy, numberOfVotes: u8) -> Result<(), Error> {}
+        /// get project policy. Called mainly by governance contract 
+        pub fn get_project_policy(&mut self, id: ProjectId) -> (ProjectPolicy, u8) {}
+        /// append bounty for specific job type. job type later can be used to create proposal for reward
+        pub fn append_bounty(&mut self, id: ProjectId, bountyType: BountyType, reward: u32) -> Result<(), Error> {}
```
[the remark this answers is not in the document]
Comment (aka-toxa) [reply applications/decentralized-security-marketplace.md:114]: So single project can have multiple bounties. 
In the case of security audits it could be different serenity levels of the issues found:

P0 - critical issue - possible money or ownership loss - 1000 USD
P1 - major issue - any bugs preventing expected business logic execution - 500 USD
P3 - minor issues - any other minor issues found - 100 USD

different project may have different set of bounties so bounty type is a formal identification of the type of work that needs to be done while project is on going state and append means that users can create their sets of bounties within their projects

File: applications/decentralized-security-marketplace.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        Completed
+    }
+    
+    pub enum ProjectPolicy {
+        AdminOnly,
+        JudgesOnly,
+        AdminAndJudges
+    }
+    
+    impl RewardsPool {
```
[the remark this answers is not in the document]
Comment (aka-toxa) [reply applications/decentralized-security-marketplace.md:106]: yes! also keep in mind the disclaimer that I put before providing this code that this is hacky and dirty pseudo code that aims to provide a bit more idea about what I have in mind

**Comment by aka-toxa** [comment]
I will push couple updates to the application in couple hours, they should address some of your comments @Noc2 @semuelle

**Comment by aka-toxa** [comment]
hey @Noc2 @semuelle 

I did a reevaluation round with my team and community CTO and we did our best to estimate all of more optimistically. The cost of development isn't changed much but it gets a bit lower. 

also I've addressed your comments about docker environment and tests in each milestone. I've tried to be more clear about intentions and steps of development

pleas take a look

**Comment by aka-toxa** [comment]
@Noc2 @semuelle hey folks

any further comments? :)

**Comment by randombishop** [comment]
Hi @aka-toxa and thank you for this grant application!

My personal independent opinion: given your Test Driven approach, I would recommend working on the first milestone and finishing the test suite before applying for a grant. That will make the deliverable specifications more concrete and easier to size and assess a grant eligibility.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. I have a few additional comments:

- “Stabilisation and production readiness” is something that we usually don’t support. Also, because we don’t know what you actually will do as part of the delivery. So I would remove this for now from the grant. If you don’t have a business model/token and you want to deploy this in the future, you could, for example, ask the treasury to fund audits and maintenance, etc.  
- The milestone tables still don’t contain the technical details, like programming/smart contract  language  
- Milestone 3 also says: “Rewards Pool implementation.”  But it's actually about the governance smart contract as far as I can see

**Comment by aka-toxa** [comment]
hey @Noc2 

> Thanks for the update. I have a few additional comments:
> 
> “Stabilisation and production readiness” is something that we usually don’t support. Also, because we don’t know what you actually will do as part of the delivery. So I would remove this for now from the grant. If you don’t have a business model/token and you want to deploy this in the future, you could, for example, ask the treasury to fund audits and maintenance, etc.

The name of milestone is misleading, it's purely about e2e tests and user guides and user manuals: how to build, test and deploy. it is not a stabilization, I've just renamed the milestone

> The milestone tables still don’t contain the technical details, like programming/smart contract language

added programming language

> Milestone 3 also says: “Rewards Pool implementation.” But it's actually about the governance smart contract as far as I can see

I don't see any mention of "rewards pool implementation" on milestone 3, can you please doublecheck?

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. I will mark the application as ready for review. But I personally won't approve it, since I think it's too expensive. Also, the last milestone doesn't contain a lot of information about the e2e test, and the first milestone is usually something we ask teams to do before they apply. But the rest of the committee might have a different opinion.

**Comment by keeganquigley** [comment]
Thanks for the application @aka-toxa a few notes:

- Typically an article is required for the last milestone. Would you be willing to write one?
- Milestone 3 seems expensive for tests, and normally we ask that these be included as part of the development. Would you be willing to remove milestone 3 and go for a level 2 grant for a PoC of the tool?

**Comment by aka-toxa** [comment]
@keeganquigley hey 

> Typically an article is required for the last milestone. Would you be willing to write one?

Yes! of course! should I add this to the application?


> Milestone 3 seems expensive for tests, and normally we ask that these be included as part of the development. Would you be willing to remove milestone 3 and go for a level 2 grant for a PoC of the tool?

yes let me move tests effort across the milestones and remove rest of it

**Comment by keeganquigley** [comment]
Thanks @aka-toxa yep, you can add this into the deliverables table under as "0e. Article"

**Review by semuelle — CHANGES_REQUESTED** [review]
Sorry for the late response, @aka-toxa. 
The application currently shows 9 person-months as a total, but the milestones only sum up to 3. 15,000 and 22,000 USD for a person-month would be indeed quite expensive. Can you clarify this?

**Comment by aka-toxa** [comment]
hey @keeganquigley an article has been added 

hey @semuelle I'm not sure I get your comment. To be honest I good only in technical stuff (design, code, lead teams). Let me introduce Ravi. He is the best person to discuss the price and be on the same page with you about estimations and pricing. I could do a horrible mistake there actually

**Comment by Rpp63** [comment]
Hello 
@aka-toxa thanks for the intro.
@semuelle, the 9 person months (PM) is a holdover from the original proposal. The current number of 3.5 PM (Milestone1 - 1 PM, Milestone2 - 1PM, Milestone3 - 1.5PM). We will make the changes.

As for the proposed budget, the current $15k/PM (~$95/hr) is the average between Smart contract engineer ($100/hr) and Full Stack dev ($85/hr). These rates are standard for MP, because 

For the purpose of this project, we can forgo most of our margin and go with $80/hr which would bring the project budget to $45,000.

Let me know if this works and Ill update the proposal

**Comment by Rpp63** [comment]
@semuelle how would you like us to proceed?

**Review by takahser — CHANGES_REQUESTED** [review]
@Rpp63 thanks for the [update](https://github.com/w3f/Grants-Program/pull/1590/commits/30c4fc64446227eef85f3830b1df8efd2ef033c1), I think that addresses @semuelle's [concerns](https://github.com/w3f/Grants-Program/pull/1590#pullrequestreview-1420149877) regarding the inconsistencies in the in-person months, though I'm not sure if writing an article would take [0.5 months](https://github.com/w3f/Grants-Program/pull/1590/commits/6e91d6b572d0c5966e24bd14385e3df9355c270d) for a fulltime employee. Anyway, it seems like the price you're planning to charge for 1 person-month is still USD 15k. Personally, I share the general sentiment already expressed by other W3F grant members here and I think that charging USD 52k for the given scope is too expensive. Hence, I'm not going to approve it neither in it's current state.

**Comment by Rpp63** [comment]
Hello @takahser 
The 0.5 months added was a correction to the clerical error (Milestone 3 should have been 1.5 months, but the application said 1 month) we had made in the initial submission, not to account for the article. No additional time, or funding, was added for the article. We can do it inhouse without any change to the requested funding. 
As for the price per engineer, I have offered updated pricing, considering the current market, which would bring the grant amount down $45,000 (~ $7k less that the last quote), if we remove all our margin we can further decrease price to $42,000. Beyond this we can reduce scope, but we're likely to reach a point of diminishing returns soon after.

**Comment by Rpp63** [comment]
@semuelle @takahser @Noc2 
Just want to touch base to see if the updated budget ($42,000) is worth exploring for you guys, so we could make a decision on this application, one way or another.

Cheers

**Comment by keeganquigley** [comment]
Hi @Rpp63 

- From what I am seeing, the budget still says $52,000 in the application. 
- A CLI is mentioned in the documentation deliverables, but not as a deliverable. Is a CLI tool going to be built?
- It's not clear how many smart contracts you will be creating. Can you elaborate?
- I know that it will be mostly tech-savvy people that use the app but would still be nice to see a front-end in Future Plans.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Just to add, reducing the price and maybe adding more technical details certainly helps with the application. Regarding the price, in general, it might make sense to bring it down to 30k for the initial grant. This way, you only need three instead of five approvals.

**Comment by Noc2** [comment]
Thanks for all the work you put into this application. However, I'm sorry to tell you that the grants committee decided to go ahead with https://github.com/w3f/Grants-Program/pull/1726 in the meantime and not support your project. Therefore, I'm closing this application. Feel free to apply for another grant, and let me know in case you have any questions.
